ADAL/src/ADAuthenticationContext+Internal.m

```diff
@@ -52,15 +52,27 @@ + (ADAuthenticationError*)errorFromDictionary:(NSDictionary*)dictionary
                                     errorCode:(ADErrorCode)errorCode
 {
     //First check for explicit OAuth2 protocol error:
-    NSString* serverOAuth2Error = [dictionary objectForKey:MSID_OAUTH2_ERROR];
+    NSString *serverOAuth2Error = [dictionary objectForKey:MSID_OAUTH2_ERROR];
     if (![NSString msidIsStringNilOrBlank:serverOAuth2Error])
     {
-        NSString* errorDetails = [dictionary objectForKey:MSID_OAUTH2_ERROR_DESCRIPTION];
+        NSString *errorDetails = [dictionary objectForKey:MSID_OAUTH2_ERROR_DESCRIPTION];
         // Error response from the server
-        NSUUID* correlationId = [dictionary objectForKey:MSID_OAUTH2_CORRELATION_ID_RESPONSE] ?
+        NSUUID *correlationId = [dictionary objectForKey:MSID_OAUTH2_CORRELATION_ID_RESPONSE] ?
                                 [[NSUUID alloc] initWithUUIDString:[dictionary objectForKey:MSID_OAUTH2_CORRELATION_ID_RESPONSE]]:
                                 nil;
-        return [ADAuthenticationError OAuthServerError:serverOAuth2Error description:errorDetails code:errorCode correlationId:correlationId];
+
+        ADErrorCode code = errorCode;
+        NSString *suberror = [dictionary objectForKey:ADAL_AUTH_SUBERROR];
+        if (suberror && [suberror isEqualToString:ADAL_AUTH_PROTECTION_POLICY_REQUIRED])
+        {
+            code = AD_ERROR_SERVER_PROTECTION_POLICY_REQUIRED;
+        }
+        
+        return [ADAuthenticationError OAuthServerError:serverOAuth2Error
+                                           description:errorDetails
+                                                  code:code
+                                         correlationId:correlationId
+                                              userInfo:suberror ? @{ADSuberrorKey:suberror} : nil];
     }
```

Comment:
Found an issue here:
Not like 2.6.4 branch, here`errorFromDictionary` is only used by auth code flow, but not refresh token flow any more.
Similar logic needs to be added to ADResponseCacheHandler.m line 52
You may need to add a new error code to MSIDError.h, and then add error mapping to ADAuthenticationErrorConverter.m (MSIDOAuthErrorDomain).